sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala

```diff
@@ -1243,19 +1243,13 @@ case class GpuMultipleContains(input: Expression, searchList: Seq[String])
   override def inputTypes: Seq[AbstractDataType] = Seq(StringType)
 
   override def doColumnar(input: GpuColumnVector): ColumnVector = {
-    assert(searchList.length > 1)
-    val accInit = withResource(Scalar.fromString(searchList.head)) { searchScalar =>
-      input.getBase.stringContains(searchScalar)
+    val targetsBytes = targets.map(t => t.getBytes).toArray
+    val boolCvs = withResource(ColumnVector.fromUTF8Strings(targetsBytes: _*)) { targetsCv =>
+      input.getBase.stringContains(targetsCv)
     }
-    searchList.tail.foldLeft(accInit) { (acc, search) =>
-      val containsSearch = withResource(Scalar.fromString(search)) { searchScalar =>
-        input.getBase.stringContains(searchScalar)
-      }
-      withResource(acc) { _ =>
-        withResource(containsSearch) { _ =>
-          acc.or(containsSearch)
-        }
-      }
+    withResource(boolCvs.tail) { _ =>
+      // boolCvs.head and intermediate values are closed within the withResource in the lambda
+      boolCvs.tail.foldLeft(boolCvs.head)((l, r) => withResource(l) { _ => l.or(r)})
```

Comment:
Still has not closed item.

```scala
val a = Array(1, 2, 5)
a.tail.foldLeft(a.head)((acc, c) => {println(s"acc is $acc, current is $c"); acc + c})
```
outputs:
```
acc is 1, current is 2
acc is 3, current is 5
```
Your approach closed 3 times, but we need to close 4 times(including the intermediate result)
Items need to close: 1, 2, 5 and 3(intermediate result)

Do the following chang:
```scala
    withResource(boolCvs.tail) { _ =>
      // boolCvs.head and intermediate values are closed within the withResource in the lambda
      boolCvs.tail.foldLeft(boolCvs.head)((l, r) => withResource(l) { _ => l.or(r)})
```
==>>
```scala
      boolCvs.tail.foldLeft(boolCvs.head)((acc, c) => 
        withResource(acc) { _ => 
          withResource(c) {_ =>
            acc.or(c)
        }
      )
```

Reply:
Can we just do a for loop. We are trying really hard to do something fancy when we can just make it simple.

```
var ret: ColumnVector = null
withResource(boolCvs) { _ =>
  boolCvs.indicies.foreach { i =>
    if (ret == null) {
      ret = boolCvs[i]
      boolCvs[i] = null
    } else {
      val tmp = ret.or(boolCvs[i])
      ret.close()
      ret = tmp
      boolCvs[i].close()
      boolCvs[i] = null
    }
  }
}

Reply:
Sorry for my [previous comment](https://github.com/NVIDIA/spark-rapids/pull/11810#discussion_r1875590543)
Haoyang's method is right,  I ignored that `withResource(boolCvs.tail)` will close multiple times if boolCvs.size > 3.

Reply:
> Can we just do a for loop. We are trying really hard to do something fancy when we can just make it simple.
> 
> ```
> var ret: ColumnVector = null
> withResource(boolCvs) { _ =>
>   boolCvs.indicies.foreach { i =>
>     if (ret == null) {
>       ret = boolCvs[i]
>       boolCvs[i] = null
>     } else {
>       val tmp = ret.or(boolCvs[i])
>       ret.close()
>       ret = tmp
>       boolCvs[i].close()
>       boolCvs[i] = null
>     }
>   }
> }
> ```
> 
> Once we have that working, then you can play around with ways to make it cleaner and/or faster.



Talked with @res-life offline, we think the following method is correct:
```scala
withResource(boolCvs.tail) { _ =>
	boolCvs.tail.foldLeft(boolCvs.head)((l, r) => withResource(l) { _ => l.or(r)})
}
```
because all elements in `boolCvs.tail` can be closed in outer `withResource`, `boolCvs.head` and intermediate results are closed in inner `withResource`.

But your approach looks better because it can close items in `boolCvs` earlier after use, so the memory usage is less. Updated to this method with some changes.

Reply:
Ok updated again, this should be cleaner and close values as early as the `foreach` method:
```scala
closeOnExcept(boolCvs.tail) { _ =>
  boolCvs.tail.foldLeft(boolCvs.head) {
    (l, r) => withResource(l) { _ =>
      withResource(r) { _ =>
        l.or(r)
      }
    }
  }
}
```
please take a look.